### ISSUE TITLE
Unexpected TypeDoesNotContainType Error after Closure use ref with conditional assignment

### ISSUE THREAD

**Issue by adamkoppede** [vimeo/psalm#10874]
In Snippet:

```php
<?php

/**
 * @var list<int> $arr
 */
$arr = [];
$hasNumberOne = false;

\usort(
    $arr, 
    static function (mixed $left, mixed $right) use (&$hasNumberOne): int {
    	if ($left === 1 || $right === 1) {
    	    $hasNumberOne = true;
            return 0;
    	}
    	return $left <=> $right;
	}
);

if ($hasNumberOne) {
    echo "hasNumberOne";
}
```

I expect `$hasNumberOne` do be of type `bool` after the call to `\usort`. Currently an [TypeDoesNotContainType](https://psalm.dev/056) error is thrown for the last if statement: https://psalm.dev/r/a41c3c69f8.

The issue can be worked around by adding an explicit type annotation with type `bool` to the first assignment of `$hasNumberOne`: https://psalm.dev/r/f23829106d

---

When I tried to look into the issue myself with an extended version of the snippet (https://psalm.dev/r/30c782d34a / [github commit for test](https://github.com/adamkoppede/psalm/commit/7ed0593bb3ee993d3bf1fe4bbc97f03c0548cef9)):

```php
<?php

/**
 * @var list<int> $arr
 */
$arr = [0, 1, 2];
$hasSomeNumber = false;
$hasNumberOne = false;

\usort(
    $arr,
    static function ($left, $right) use (&$hasSomeNumber, &$hasNumberOne): int {
        $hasSomeNumber = true;
    	if ($left === 1 || $right === 1) {
    	    $hasNumberOne = true;
            return 0;
    	}
    	return $left <=> $right;
	}
);

if ($hasSomeNumber) { // has expected type `bool`
    echo "hasSomeNumber";
}

if ($hasNumberOne) { // has unexpected type `false`
    echo "hasNumberOne";
}
```

I found that the change from `TFalse` to `TTrue` inside the conditional is correctly determined in the local variable `$if_context`  in [`\Psalm\Internal\Analyzer\Statements\Block\IfElse\IfAnalyzer::analyze`](https://github.com/vimeo/psalm/blob/5.x/src/Psalm/Internal/Analyzer/Statements/Block/IfElse/IfAnalyzer.php#L52). However, it isn't carried up into `$ref_context` in [`\Psalm\Internal\Analyzer\FunctionLikeAnalyzer::analyze`](https://github.com/vimeo/psalm/blob/5.x/src/Psalm/Internal/Analyzer/FunctionLikeAnalyzer.php#L365-L386). There in `$ref_context`, `$hasSomeNumber` is of expected type `TBool` while `$hasNumberOne` remains of unexpected type `TFalse`.
